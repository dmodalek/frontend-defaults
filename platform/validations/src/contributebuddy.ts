import { ContributebuddyAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import {
	ValidationResult,
	ValidationSeverityLevel,
	validateInstallation,
} from '@namics/frontend-defaults-platform-core';

export const contributebuddyValidation = async (
	cwd: string,
	analytics: ContributebuddyAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.contributebuddy) {
		// contributebuddy is used in the project, validation below
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.contributebuddyInstallation,
				name: 'Contributebuddy',
				source: 'contributebuddyValidation',
			})
		);

		if (!analytics.contributebuddyConfigFolder) {
			validationResults.push({
				level: ValidationSeverityLevel.error,
				message: 'No .contributebuddy folder found, make sure you set-up correctly',
				source: 'contributebuddyValidation',
			});
		}

		return validationResults;
	}

	return [
		{
			level: ValidationSeverityLevel.info,
			message: 'We recommend to use contribute-buddy to obtain maintainability',
			source: 'contributebuddyValidation',
		},
	];
};
