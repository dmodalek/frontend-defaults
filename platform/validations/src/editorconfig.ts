import { EditorconfigAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const editorconfigValidation = async (cwd: string, analytics: EditorconfigAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.editorConfig) {
		// TODO: editorconfig is available, so we need to validate all things
		return [];
	}

	return [{
		level: ValidationSeverityLevel.error,
		message: 'Default namics editorconfig file is required in every project',
		source: 'editorconfigValidation'
	}];
}