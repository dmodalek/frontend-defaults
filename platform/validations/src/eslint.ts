import { ESLintAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel, validateInstallation } from '@namics/frontend-defaults-platform-core';

/**
 * TODO: Should we check if a project is a TS project in advance? So we wouldn't need to run this validation atm.
 * I think we'll need to update this as soon as TSLint will be integrated into ESLint
 */
export const eslintValidation = async (cwd: string, analytics: ESLintAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.eslint) {
		// TODO: eslint is available, so we need to validate all things
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.eslintInstallation,
				name: 'ESLint',
				source: 'eslintValidation'
			})
		);

		if (!analytics.eslintIgnore) {
			validationResults.push({
				level: ValidationSeverityLevel.warning,
				message: 'No .eslintignore found',
				source: 'eslintValidation'
			})
		}

		return validationResults;
	}

	return [{
		level: ValidationSeverityLevel.warning,
		message: 'ESLint configuration missing',
		source: 'eslintValidation'
	}];
}