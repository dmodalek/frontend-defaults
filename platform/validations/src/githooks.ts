import { GitHooksAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel, validateInstallation } from '@namics/frontend-defaults-platform-core';

export const gitHooksValidation = async (cwd: string, analytics: GitHooksAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.gitHooks) {
		let validationResults: ValidationResult[] = [];
		validationResults = validationResults.concat(
			validateInstallation({
				installation: analytics.gitHooksInstallation,
				name: 'GitHooks',
				source: 'gitHooksValidation'
			})
		);

		if (analytics.gitHooksLocation === 'rc') {
			validationResults.push({
				level: ValidationSeverityLevel.warning,
				message: 'For future projects we decided to use the package.json for configuration (.rc file used)',
				source: 'gitHooksValidation'
			})
		}

		return validationResults;
	}

	return [{
		level: ValidationSeverityLevel.info,
		message: 'GitHooks are missing',
		source: 'gitHooksValidation'
	}];
}