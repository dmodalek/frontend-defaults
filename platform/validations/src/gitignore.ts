import { GitIgnoreAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const gitIgnoreValidation = async (
	cwd: string,
	analytics: GitIgnoreAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.gitignore) {
		// TODO: gitignore is available, so we need to validate the contents
		return [];
	}

	return [
		{
			level: ValidationSeverityLevel.error,
			message: 'Git ignore missing',
			source: 'gitIgnoreValidation',
		},
	];
};
