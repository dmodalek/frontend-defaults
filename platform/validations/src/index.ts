export * from './editorconfig';
export * from './eslint';
export * from './githooks';
export * from './gitignore';
export * from './license';
export * from './nitro';
export * from './node';
export * from '../../patches/src/npmrc';
export * from './readme';
export * from './tslint';
export * from './typescript';
export * from './webpack';
