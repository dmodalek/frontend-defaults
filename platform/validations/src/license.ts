import { LicenseAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const licenseValidation = async (cwd: string, analytics: LicenseAnalyzerResult): Promise<ValidationResult[]> => {
	if (analytics.license) {
		// TODO: license is available, so we need to validate the content
		return [];
	}

	return [{
		level: ValidationSeverityLevel.warning,
		message: 'You don\'t have a license file in your project',
		source: 'licenseValidation'
	}];
}