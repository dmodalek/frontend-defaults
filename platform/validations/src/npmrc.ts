import { NPMRCAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const npmrcValidation = async (cwd: string, analytics: NPMRCAnalyzerResult): Promise<ValidationResult[]> => {
    const validationResults: ValidationResult[] = [];

    if (!analytics.npmrc) {
        // No .npmrc file found
        validationResults.push({
            message: `No .npmrc file found to reference to the correct node version`,
            source: 'NPMRCValidation',
            patch: [['NPMRCPatch', { create: true }]],
            level: ValidationSeverityLevel.error,
        });
    }

    if (analytics.npmrc && !analytics.npmrcSaveExactEnabled) {
        // .npmrc file found but save-exact not enabled
        validationResults.push({
            message: `Save exact is not enabled in your .npmrc file`,
            source: 'NPMRCValidation',
            patch: ['NPMRCPatch'],
            level: ValidationSeverityLevel.warning,
        });
    }

    return validationResults;
}