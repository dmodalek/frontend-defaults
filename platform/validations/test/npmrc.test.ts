import { ProjectValidator, ProjectAnalyzer, ValidationExceptionLevel } from '@namics/frontend-defaults-platform-core';
import { getFixtureDir, checkInternetConnection } from './utils';
import { NPMRCValidation } from '../src/npmrc';
import { NPMRCAnalyzerResult, NPMRCAnalyzer } from '@namics/frontend-defaults-platform-analyzers';

const getAnalyzerFor = async (fixtureContext: string): Promise<ProjectAnalyzer<NPMRCAnalyzerResult>> => {
	return new ProjectAnalyzer<NPMRCAnalyzerResult>({
		context: fixtureContext,
		analyzers: [NPMRCAnalyzer],
	}).boot();
};

beforeAll(async () => {
	if (!(await checkInternetConnection())) {
		console.debug(`Please only run tests while connected to the internet!`);
		process.exit(1); // kill child worker
	}
});

describe('Validations', () => {
	describe('NPMRCValidation', () => {
		it('should be creatable without errors', () => {
			expect(async () => {
				const analyzer = await getAnalyzerFor(getFixtureDir('npmrc-project'));
				new ProjectValidator({
					analyzer,
					validations: [NPMRCValidation],
				});
			}).not.toThrow();
		});

		it('should recognize no issues if everything is valid', async () => {
			const analyzer = await getAnalyzerFor(getFixtureDir('npmrc-project'));
			const validator = new ProjectValidator({
				analyzer,
				validations: [NPMRCValidation],
			});

			const patches = await validator.validate();
			expect(patches).toHaveLength(0);
			expect(validator.validation).toHaveLength(0);
		});

		it('should recognize a missing npmrc', async () => {
			const analyzer = await getAnalyzerFor(getFixtureDir('default-project'));
			const validator = new ProjectValidator({
				analyzer,
				validations: [NPMRCValidation],
			});

			const patches = await validator.validate();
			expect(patches).toHaveLength(1);
			expect(validator.validation).toHaveLength(1);
			expect(patches[0].patch).toEqual('CreateNPMRCPatch');
			expect(validator.validation[0].toString()).toMatchInlineSnapshot(
				`"Validate NPMRCValidation: No .npmrc file found to reference to the correct node version [ERROR]"`
			);
		});

		it('should detect the missing save-exact flag', async () => {
			const analyzer = await getAnalyzerFor(getFixtureDir('npmrc-project-no-save-exact'));
			const validator = new ProjectValidator({
				analyzer,
				validations: [NPMRCValidation],
			});

			const patches = await validator.validate();
			expect(patches).toHaveLength(1);
			expect(validator.validation).toHaveLength(1);
			expect(patches[0].patch).toEqual('AddSaveExactToNPMRCPatch');
			expect(validator.validation[0].toString()).toMatchInlineSnapshot(
				`"Validate NPMRCValidation: Save exact is not enabled in your .npmrc file [WARNING]"`
			);
		});
	});
});
