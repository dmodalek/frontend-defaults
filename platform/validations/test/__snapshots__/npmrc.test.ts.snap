// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validations NPMRCValidation should detect the missing save-exact flag 2`] = `
Array [
  Object {
    "level": "WARNING",
    "message": "Save exact is not enabled in your .npmrc file",
    "patch": Array [
      "NPMRCPatch",
    ],
    "source": "NPMRCValidation",
  },
]
`;

exports[`Validations NPMRCValidation should recognize a missing npmrc 2`] = `
Array [
  Object {
    "level": "ERROR",
    "message": "No .npmrc file found to reference to the correct node version",
    "patch": Array [
      Array [
        "NPMRCPatch",
        Object {
          "create": true,
        },
      ],
    ],
    "source": "NPMRCValidation",
  },
]
`;

exports[`Validations NPMRCValidation should recognize no issues if everything is valid 1`] = `Array []`;
